### PR TITLE
🐛  fix unhandled errors

### PR DESCRIPTION
**Pull request by sbueringer**
Signed-off-by: Stefan Büringer buringerst@vmware.com

**What this PR does / why we need it**:
I think it makes sense to either handle errors or ignore them explicitly

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Implements a part of #4622

### CODE REVIEW

File: cmd/clusterctl/cmd/config_provider.go [hunk 1]
```diff
@@ -187,8 +187,10 @@ func printComponents(c client.Components, output string) error {
 		if _, err := os.Stdout.Write(yaml); err != nil {
 			return errors.Wrap(err, "failed to write yaml to Stdout")
 		}
-		os.Stdout.WriteString("\n")
-		return err
+		if _, err := os.Stdout.WriteString("\n"); err != nil {
+			return errors.Wrap(err, "failed to write yaml to Stdout")
```
Comment (JoelSpeed) [inline cmd/clusterctl/cmd/config_provider.go:191]: Nit, while this is unlikely to ever actually happen, the error on this line is identical to 188, that could make it slightly harder to debug if you just end up with an error and not a line number the error was generated from. Might be better to change this slightly so it is a unique error, WDYT?

File: controllers/remote/cluster_cache_healthcheck_test.go [hunk 2]
```diff
@@ -168,7 +168,8 @@ func TestClusterCacheHealthCheck(t *testing.T) {
 			g.Expect(err).NotTo(HaveOccurred())
 			l, err := net.ListenTCP("tcp", addr)
 			g.Expect(err).NotTo(HaveOccurred())
-			l.Close()
+			err = l.Close()
+			g.Expect(err).NotTo(HaveOccurred())
```
Comment (JoelSpeed) [inline controllers/remote/cluster_cache_healthcheck_test.go:172]: Could inline this as it only returns the one variable, eg
```suggestion
			g.Expect(l.Close()).To(Succeed())
```

File: controllers/mdutil/util.go [hunk 3]
```diff
@@ -701,7 +701,7 @@ func DeepHashObject(hasher hash.Hash, objectToWrite interface{}) {
 		DisableMethods: true,
 		SpewKeys:       true,
 	}
-	printer.Fprintf(hasher, "%#v", objectToWrite)
```
Comment (enxebre) [inline controllers/mdutil/util.go:704]: why are we ignoring this error?

**Review by JoelSpeed — COMMENTED** [review]
This is fine as is but added a couple of notes for potential improvements, very minor/stylistic and definitely non-blocking

File: controllers/mdutil/util.go (hunk 3 above)
Comment (sbueringer) [reply controllers/mdutil/util.go:704]: Good question, we probably shouldn't

File: cmd/clusterctl/cmd/config_provider.go (hunk 1 above)
Comment (sbueringer) [reply cmd/clusterctl/cmd/config_provider.go:191]: Yup, will do

File: controllers/mdutil/util.go (hunk 3 above)
Comment (sbueringer) [reply controllers/mdutil/util.go:704]: Implemented err handling now

**Comment by sbueringer** [comment]
Thx for the reviews, ptal @enxebre @JoelSpeed 

(I hope/assume the apidiff is not a problem)

**Review by JoelSpeed — COMMENTED** [review]
/lgtm

Looking at the diff, seems we've changed the signature of some machinedeployment controlelr utils, that appears to be handled within this PR and I doubt that code is used outside of the project, so I think that should be ok

**Comment by sbueringer** [comment]
/test pull-cluster-api-test-main
I assume those were flakes, tests worked locally.

File: controllers/mdutil/util.go [hunk 4]
```diff
@@ -693,22 +693,28 @@ func CloneSelectorAndAddLabel(selector *metav1.LabelSelector, labelKey, labelVal
...
 	hasher.Reset()
 	printer := spew.ConfigState{
 		Indent:         " ",
 		SortKeys:       true,
 		DisableMethods: true,
 		SpewKeys:       true,
 	}
-	printer.Fprintf(hasher, "%#v", objectToWrite)
+	_, err := printer.Fprintf(hasher, "%#v", objectToWrite)
+	if err != nil {
```
Comment (vincepri) [inline controllers/mdutil/util.go:705]:
```suggestion
	if _, err := printer.Fprintf(hasher, "%#v", objectToWrite); err != nil {
```

File: controllers/mdutil/util.go [hunk 5]
```diff
@@ -693,22 +693,28 @@ func CloneSelectorAndAddLabel(selector *metav1.LabelSelector, labelKey, labelVal
 // DeepHashObject writes specified object to hash using the spew library
 // which follows pointers and prints actual values of the nested objects
 // ensuring the hash does not change when a pointer changes.
-func DeepHashObject(hasher hash.Hash, objectToWrite interface{}) {
+func DeepHashObject(hasher hash.Hash, objectToWrite interface{}) error {
```
Comment (vincepri) [inline controllers/mdutil/util.go:696]: This breaks API compatibility, we should deprecate this method and add a new one instead that returns an error, for example it'd also be great to rename it as `SpewHashObject` to signal that we're using the Spew library to hash objects.

/hold

**Comment by sbueringer** [comment]
@vincepri I also added a `ComputeSpewHash` to avoid breaking API compatiblity of `ComputeHash`. PTAL

**Review by vincepri — COMMENTED** [review]
/lgtm
/hold cancel

**Review by vincepri — APPROVED** [review]
/approve

**Comment by vincepri** [comment]
This is a new, @sbueringer can you check if the flake is reproducible locally?

/hold

**Comment by sbueringer** [comment]
@vincepri I don't think it's related:
* the previous failure had other failed tests (the same commit has 2 successfull runs of this test): https://prow.k8s.io/pr-history/?org=kubernetes-sigs&repo=cluster-api&pr=4895
* it already failed on the main branch last Friday: 
    * https://testgrid.k8s.io/sig-cluster-lifecycle-cluster-api#capi-test-main&include-filter-by-regex=TestGetWorkloadCluster&exclude-non-failed-tests=
    * https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/periodic-cluster-api-test-main/1413261062446256128
* Ran it locally a few times without error
* As far as I can see I didn't change any code related to the kubeadm control plane controller

Hopefully the retest will also work :)

/retest

**Comment by sbueringer** [comment]
@vincepri I think we can merge this PR.

**Pull request merged**